sm/tedge_agent/src/agent.rs

```diff
@@ -345,29 +367,130 @@ impl SmAgent {
         Ok(())
     }
 
-    async fn fail_pending_operation(&self, mqtt: &Client) -> Result<(), AgentError> {
+    async fn handle_restart_operation(
```

Comment:
This function has far too many responsibilities.

* Decoding the request message
* Managing the agent state
* Forwarding the operation status to the requester
* And doing the restart.

This should be at least be broken in more specific functions. More importantly error handling must be done *around* all these steps to ensure that any error is properly handled.